internal/api/mfa.go

```diff
@@ -399,17 +388,49 @@ func (a *API) ChallengeFactor(w http.ResponseWriter, r *http.Request) error {
 	})
 }
 
-func (a *API) verifyPhoneFactor(w http.ResponseWriter, r *http.Request, params *VerifyFactorParams) error {
+func (a *API) ChallengeFactor(w http.ResponseWriter, r *http.Request) error {
 	ctx := r.Context()
 	config := a.config
+	factor := getFactor(ctx)
+	user := getUser(ctx)
+
+	if !factor.IsOwnedBy(user) {
+		return notFoundError(ErrorCodeMFAFactorNotFound, "MFA factor not found")
+	}
+
+	switch factor.FactorType {
+	case models.Phone:
+		if !config.MFA.Phone.VerifyEnabled {
+			return unprocessableEntityError(ErrorCodeMFAPhoneEnrollDisabled, "MFA verification is disabled for Phone")
+		}
+		return a.challengePhoneFactor(w, r)
+
+	case models.TOTP:
+		// Prior to the introduction of MFA.TOTP.VerifyEnabled,
+		// MFA.Enabled was used to configure whether TOTP was on. So
+		// both have to be set to false to regard the feature as
+		// disabled.
+		if !config.MFA.Enabled && !config.MFA.TOTP.VerifyEnabled {
+			return unprocessableEntityError(ErrorCodeMFATOTPEnrollDisabled, "MFA verification is disabled for TOTP")
+		}
+		return a.challengeTOTPFactor(w, r)
+	default:
+		return badRequestError(ErrorCodeValidationFailed, "factor_type needs to be TOTP or Phone")
+	}
+
+}
+
+func (a *API) verifyTOTPFactor(w http.ResponseWriter, r *http.Request, params *VerifyFactorParams) error {
+	var err error
+	ctx := r.Context()
 	user := getUser(ctx)
 	factor := getFactor(ctx)
+	config := a.config
 	db := a.db.WithContext(ctx)
 	currentIP := utilities.GetIPAddress(r)
 
 	if !factor.IsOwnedBy(user) {
-		return notFoundError(ErrorCodeMFAFactorNotFound, "MFA factor not found")
-
+		return internalServerError(InvalidFactorOwnerErrorMessage)
```

Comment:
kinda confused here - is it supposed to return a not found or internal server error? currently, we return a mix of them in this PR

also, it might be worth moving this check to `loadFactor` instead to simplify things

Reply:
We return `internalServerError` on the TOTP methods so as not to change the status code and preserve backward compatibility

We return `notFoundError` on the phone factor as it's new and there's no existing concern w.r.t backward compatibility.

Ideally we move TOTP factors to use `notFoundError` in the future as well

Reply:
Added a comment to clarify why it's `internalServerError`